coalib/core/DependencyTracker.py

```diff
@@ -0,0 +1,162 @@
+from coalib.core.Graphs import traverse_graph
+
+
+class DependencyTracker:
+    """
+    A ``DependencyTracker`` allows to register and manage dependencies between
+    objects.
+
+    This class uses a directed graph to track relations.
+
+    Add a dependency relation between two objects:
+
+    >>> object1 = object()
+    >>> object2 = object()
+    >>> tracker = DependencyTracker()
+    >>> tracker.add(object2, object1)
+
+    This would define that ``object1`` is dependent on ``object2``.
+
+    If you define that ``object2`` has its dependency duty fulfilled, you can
+    resolve it:
+
+    >>> resolved = tracker.resolve(object2)
+    >>> resolved
+    {<object object at ...>}
+    >>> resolved_object = resolved.pop()
+    >>> resolved_object is object1
+    True
+
+    This returns all objects that are now freed, meaning they have no
+    dependencies any more.
+
+    >>> object3 = object()
+    >>> tracker.add(object2, object1)
+    >>> tracker.add(object3, object1)
+    >>> tracker.resolve(object2)
+    set()
+    >>> tracker.resolve(object3)
+    {<object object at ...>}
+    """
+
+    def __init__(self):
+        self._dependency_dict = {}
+
+    def add(self, dependency, dependant):
+        """
+        Add a dependency relation.
+
+        This function does not check for circular dependencies.
+
+        >>> tracker = DependencyTracker()
+        >>> tracker.add(0, 1)
+        >>> tracker.add(0, 2)
+        >>> tracker.resolve(0)
+        {1, 2}
+
+        :param dependency:
+            The object that is the dependency.
+        :param dependant:
+            The object that is the dependant.
+        """
+        if dependency not in self._dependency_dict:
+            self._dependency_dict[dependency] = set()
```

Comment:
The problem with defaultdicts that accesses do create new sets and those get placed into the dict. That makes it non-empty any more and `bool(dc)` would then yield `True` though there are effectively no dependency relations. I rely on that behaviour :+1: